File: crc32.go
// crc32.go
package main

import (
	"bufio"
	"encoding/binary"
	"flag"
	"fmt"
	"hash/crc32"
	"io"
	"os"
	"strconv"
	"strings"
)

func computeCRC32(data []byte, useCRC32C bool) uint32 {
	if useCRC32C {
		return crc32.Checksum(data, crc32.MakeTable(crc32.Castagnoli))
	}
	return crc32.ChecksumIEEE(data)
}

func computeFileCRC32(filename string, useCRC32C bool) (uint32, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, err
	}
	defer file.Close()
	var table *crc32.Table
	if useCRC32C {
		table = crc32.MakeTable(crc32.Castagnoli)
	} else {
		table = crc32.IEEETable
	}
	hash := crc32.New(table)
	if _, err := io.Copy(hash, file); err != nil {
		return 0, err
	}
	return hash.Sum32(), nil
}

func formatHash(val uint32, format string) string {
	switch format {
	case "hex":
		return fmt.Sprintf("0x%08X", val)
	case "dec":
		return fmt.Sprintf("%d", val)
	case "bin":
		return fmt.Sprintf("%032b", val)
	default:
		return fmt.Sprintf("0x%08X", val)
	}
}

func main() {
	var (
		useCRC32C bool
		check     string
		dec       bool
		bin       bool
		progress  bool
	)
	flag.BoolVar(&useCRC32C, "crc32c", false, "Использовать CRC32C")
	flag.StringVar(&check, "check", "", "Сравнить с указанным хэшем (HEX)")
	flag.BoolVar(&dec, "dec", false, "Вывод в десятичном формате")
	flag.BoolVar(&bin, "bin", false, "Вывод в бинарном формате")
	flag.BoolVar(&progress, "progress", false, "Показывать прогресс")
	flag.Parse()
	args := flag.Args()

	format := "hex"
	if dec {
		format = "dec"
	} else if bin {
		format = "bin"
	}

	// Если нет аргументов и есть данные в stdin
	if len(args) == 0 {
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			data, err := io.ReadAll(os.Stdin)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Ошибка чтения stdin: %v\n", err)
				os.Exit(1)
			}
			hashVal := computeCRC32(data, useCRC32C)
			fmt.Println(formatHash(hashVal, format))
			return
		}
		fmt.Println("Не указаны данные. Используйте: crc32 <строка или файл>")
		os.Exit(1)
	}

	for _, item := range args {
		info, err := os.Stat(item)
		if err == nil && !info.IsDir() {
			// Это файл
			hashVal, err := computeFileCRC32(item, useCRC32C)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Ошибка чтения %s: %v\n", item, err)
				continue
			}
			output := fmt.Sprintf("%s: %s", item, formatHash(hashVal, format))
			if check != "" {
				expected, _ := strconv.ParseUint(strings.TrimPrefix(check, "0x"), 16, 32)
				status := "✅ OK"
				if uint32(expected) != hashVal {
					status = "❌ FAIL"
				}
				output += fmt.Sprintf(" (check: %s)", status)
			}
			fmt.Println(output)
		} else {
			// Строка
			data := []byte(item)
			hashVal := computeCRC32(data, useCRC32C)
			output := formatHash(hashVal, format)
			if check != "" {
				expected, _ := strconv.ParseUint(strings.TrimPrefix(check, "0x"), 16, 32)
				status := "✅ OK"
				if uint32(expected) != hashVal {
					status = "❌ FAIL"
				}
				output += fmt.Sprintf(" (check: %s)", status)
			}
			fmt.Println(output)
		}
	}
}

File: crc32.js
// crc32.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const yargs = require('yargs/yargs');
const { hideBin } = require('yargs/helpers');
const crc32 = require('crc/crc32');
const crc32c = require('crc/crc32c');

function computeCRC32(data, useCRC32c = false) {
    if (useCRC32c) {
        return crc32c(data).toString(16).toUpperCase();
    } else {
        return crc32(data).toString(16).toUpperCase();
    }
}

function computeFileCRC32(filename, useCRC32c = false) {
    return new Promise((resolve, reject) => {
        const stream = fs.createReadStream(filename);
        const crc = useCRC32c ? crc32c : crc32;
        let hash = 0;
        stream.on('data', (chunk) => {
            hash = crc(chunk, hash);
        });
        stream.on('end', () => {
            resolve(hash.toString(16).toUpperCase());
        });
        stream.on('error', reject);
    });
}

function formatHash(hash, format) {
    if (format === 'hex') {
        return `0x${hash.padStart(8, '0')}`;
    } else if (format === 'dec') {
        return parseInt(hash, 16).toString();
    } else if (format === 'bin') {
        return parseInt(hash, 16).toString(2).padStart(32, '0');
    }
    return `0x${hash.padStart(8, '0')}`;
}

async function main() {
    const argv = yargs(hideBin(process.argv))
        .usage('Использование: $0 <строка или файл> [опции]')
        .option('crc32c', { type: 'boolean', description: 'Использовать CRC32C' })
        .option('check', { type: 'string', description: 'Сравнить с хэшем (HEX)' })
        .option('dec', { type: 'boolean', description: 'Вывод в десятичном формате' })
        .option('bin', { type: 'boolean', description: 'Вывод в бинарном формате' })
        .help()
        .parse();

    const inputs = argv._;
    const useCRC32c = argv.crc32c || false;
    const check = argv.check || null;
    let format = 'hex';
    if (argv.dec) format = 'dec';
    if (argv.bin) format = 'bin';

    // Проверка stdin
    if (inputs.length === 0 && !process.stdin.isTTY) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });
        let data = '';
        for await (const line of rl) {
            data += line + '\n';
        }
        const hash = computeCRC32(Buffer.from(data), useCRC32c);
        console.log(formatHash(hash, format));
        return;
    }

    if (inputs.length === 0) {
        console.error('Не указаны данные.');
        process.exit(1);
    }

    for (const item of inputs) {
        if (fs.existsSync(item) && fs.statSync(item).isFile()) {
            const hash = await computeFileCRC32(item, useCRC32c);
            let output = `${item}: ${formatHash(hash, format)}`;
            if (check) {
                const expected = parseInt(check.startsWith('0x') ? check.slice(2) : check, 16);
                const actual = parseInt(hash, 16);
                output += (expected === actual) ? ' (✅ OK)' : ' (❌ FAIL)';
            }
            console.log(output);
        } else {
            const data = Buffer.from(item, 'utf8');
            const hash = computeCRC32(data, useCRC32c);
            let output = formatHash(hash, format);
            if (check) {
                const expected = parseInt(check.startsWith('0x') ? check.slice(2) : check, 16);
                const actual = parseInt(hash, 16);
                output += (expected === actual) ? ' (✅ OK)' : ' (❌ FAIL)';
            }
            console.log(output);
        }
    }
}

main().catch(console.error);
